### PR TITLE
[FLINK-19681][checkpointing] Timeout aligned checkpoints

### PR DESCRIPTION
**Pull request by rkhachatryan**
This is an updated version of #13827 to pass the tests and address the remaining feedback.

### CODE REVIEW

**Comment by rkhachatryan** [comment]
@flinkbot run azure

**Comment by SteNicholas** [comment]
@rkhachatryan Did you complete working for FLINK-19681? I'm waiting for you for the completion of this pull request.

**Comment by rkhachatryan** [comment]
@SteNicholas yes, but I think the PR won't be merged soon. I expect a lot of feedback from @pnowojski

The latest AZP failure is caused by FLINK-19863

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java [hunk 1]
```diff
@@ -193,6 +202,11 @@ private CheckpointBarrierBehaviourController chooseController(CheckpointBarrier
 
 	private boolean canTimeout(CheckpointBarrier barrier) {
 		return barrier.getCheckpointOptions().isTimeoutable() &&
-			barrier.getCheckpointOptions().getAlignmentTimeout() < (System.currentTimeMillis() - barrier.getTimestamp());
+			barrier.getId() <= lastSeenBarrier &&
+			barrier.getCheckpointOptions().getAlignmentTimeout() * 1_000_000 < (System.nanoTime() - firstBarrierArrivalTime);
+	}
+
+	private long getArrivalTime(CheckpointBarrier announcedBarrier) {
+		return announcedBarrier.getCheckpointOptions().isTimeoutable() ? System.nanoTime() : Long.MAX_VALUE;
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java]: I'm not entirely convinced if this a better approach. 

For me, using global checkpoint time (in other words recently added checkpoinStartDelay metric) seemed easier to understand for the user. If aligned checkpoint barrier is taking too much time to reach given tasks - we are timeouting it to unaligned checkpoint that overtakes in-flight data. This seemed easier to comprehend and easier to explain  compared to time outing alignment on some subtask?

Secondly your proposed change will not work with single input tasks without active timeouts?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java [hunk 2]
```diff
@@ -114,6 +114,7 @@ public void barrierAnnouncement(
 			lastSeenBarrier = barrier.getId();
 			firstBarrierArrivalTime = getArrivalTime(barrier);
 		}
+		activeController = chooseController(barrier);
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java:117]: ? 

Why is it not enough to do this in `preProcessFirstBarrierOrAnnouncement`?

(as in other places: a unit test would be helpful)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java [hunk 3]
```diff
@@ -146,28 +146,7 @@ private void switchToUnaligned(
...
-			return Optional.empty();
-		}
-		else {
-			checkState(activeController == alignedController);
-			Optional<CheckpointBarrier> triggerResult = activeController.postProcessLastBarrier(
-				channelInfo,
-				barrier);
-			checkState(triggerResult.isPresent());
-			return triggerResult;
-		}
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java]: Why have you removed this code? What was the problem?

Was it subsumed by `switchToUnaligned ` call happening in the last `barrierReceived` call?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/CheckpointBarrier.java [hunk 4]
```diff
@@ -119,6 +119,6 @@ public boolean isCheckpoint() {
 	}
 
 	public CheckpointBarrier asUnaligned() {
-		return checkpointOptions.isUnalignedCheckpoint() ? this : new CheckpointBarrier(getId(), getTimestamp(), getCheckpointOptions().asTimedOut());
+		return checkpointOptions.isUnalignedCheckpoint() ? this : new CheckpointBarrier(getId(), getTimestamp(), getCheckpointOptions().toUnaligned());
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/CheckpointBarrier.java:122]: part of a previous fixup?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 5]
```diff
@@ -569,14 +569,23 @@ public void convertToPriorityEvent(int sequenceNumber) throws IOException {
...
+			// don't convert the barrier itself (barrier controller might not have been switched yet)
+			AbstractEvent e = EventSerializer.fromBuffer(toPrioritize.buffer, this.getClass().getClassLoader());
+			toPrioritize.buffer.setReaderIndex(0);
+			toPrioritize = new SequenceBuffer(EventSerializer.toBuffer(e, true), toPrioritize.sequenceNumber);
 			firstPriorityEvent = addPriorityBuffer(toPrioritize); 	// note that only position of the element is changed
 																	// converting the event itself would require switching the controller sooner
 		}
 		if (firstPriorityEvent) {
-			notifyPriorityEvent(sequenceNumber);
+			notifyPriorityEventForce(); // use force here because the barrier SQN might be seen by gate during the announcement
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:581]: nit: I don't understand this comment, can you rephrase/elaborate a bit?

**Review by pnowojski — COMMENTED** [review]
Thanks for taking care of debugging and fixing my code. I presume the tests are stable now? Even if you enable timeout by default?

I've left a couple of comments, most are questions.

One big thing that is missing are the unit tests for the fixes that you implemented. Given how difficult was it to track down those bugs relaying on UCITCase, I would say unit tests are must have.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 2 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java:117]: Consider a checkpoint with two barriers from two channels: UC and AC:
1. Process AC announcement. Set `SingleCheckpointBarrierHandler.currentCheckpointId`. Don't timeout. 
2. Process UC barrier (no announcements). Because `currentCheckpointId` was set `checkSubsumedCheckpoint` does nothing, i.e. `preProcessFirstBarrierOrAnnouncement` not called and controller so `activeController` == AC
3. Further process (UC) barrier - now AC controller is processing UC barrier

> (as in other places: a unit test would be helpful)

I agree, will add it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java]: >  easier to understand

I agree that it might be true for some users, but not for all. During the [previous discussion](https://github.com/apache/flink/pull/13827#discussion_r527794600), and also the one before, the consensus was that it's **not** easier to understand. However, we can discuss it again. 
(also there are some more technical advantages of "local" timeouts)

> Secondly your proposed change will not work with single input tasks without active timeouts?

Why, could you explain?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 3 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java]: Yes, with `switchToUnaligned` in `barrierReceived` it seems unnecessary.
With this code, `UnalignedCheckpointITCase` fails: some tests on `checkState` (line 164 above), some hanging up.

Besides, why timeout alignment if it's the last barrier? This essentially means that alignment is done.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/CheckpointBarrier.java (hunk 4 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/CheckpointBarrier.java:122]: No, I think it belongs to the right commit. With `CheckpointOptions` hotfix/refactoring, this method converts it to `unaligned`, which can't be timed out.

**Comment by rkhachatryan** [comment]
Thanks for reviewing @pnowojski.
Yes, the tests are stable (I'll check it again though since I rebased the PR recently).
I agree about unit tests, will add them and address other issues.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java]: > Why, could you explain?

On a second thought, maybe it will partially, but not fully as well.  guess we still have the code, that we can timeout to UC on the last processed barrier? So in case of single channel:
1. announcement is processed (first announcement will never timeout in this version) it won't timeout
2. barrier will be processed, and only it can timeout

So a timeout on the first network exchange will work worse. That's a bit problematic, especially for simple jobs, with for example just a single exchange. Previous version would cut the checkpointing time by half, this version will do worse than that, in a way that's hard to quantify for me. 

There is some extreme corner case when imagine there is a heavy back pressure, but all CB are processed at the same time. That means announcements in this version wouldn't cause timeout (it would in my older proposal), and this version will need to wait for some CB to be processed (which can take long time).

Active timeout would alleviate this problem though.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 2 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java:117]: But doesn't it mean we should support time outing on every UC barrier? And it looks like we are doing that:
```
	@Override
	public Optional<CheckpointBarrier> barrierReceived(InputChannelInfo channelInfo, CheckpointBarrier barrier) throws IOException, CheckpointException {
		if (barrier.getCheckpointOptions().isUnalignedCheckpoint() && activeController == alignedController) {
			barrier = barrier.asUnaligned();
			switchToUnaligned(channelInfo, barrier);
			activeController.barrierReceived(channelInfo, barrier);
			return Optional.of(barrier);
		}
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 3 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java]: > Besides, why timeout alignment if it's the last barrier? This essentially means that alignment is done.

The alignment is done, but that's just input. As we do not have code to timeout outputs, it's better to timeout to UC even if the alignment was completed (although too late). Think especially about a case with just a single input channel.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java]: With a single input, this is not an alignment issue but a (related) back-pressure issue. And I think this is the right way to solve it: 
> Active timeout would alleviate this problem though.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 2 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java:117]: But there is a `preProcessFirstBarrier` before the switch in `barrierReceived`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 3 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java]: Isn't it enough to just convert the barrier upon triggering checkpoint in subtask?

This is what I've done in `[FLINK-19681][checkpointing] Use converted barrier after disabling alignment`. It is needed anyways for correct handling in `SubtaskCheckpointCoordinatorImpl`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlignedController.java [hunk 6]
```diff
@@ -120,6 +122,12 @@ public void obsoleteBarrierReceived(
 		resumeConsumption(channelInfo);
 	}
 
+	protected void resetPendingCheckpoint(long cancelledId) {
+		for (final CheckpointableInput input : inputs) {
+			input.checkpointStopped(cancelledId);
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlignedController.java:131]: I've added `AlternatingControllerTest.testChannelResetOnNewBarrier`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 2 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java:117]: I've added `AlternatingControllerTest.testSwitchToUnalignedByUpstream`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java (hunk 5 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: Rephrased as:
```
// forcibly notify about the priority event
// instead of passing barrier SQN to be checked
// because this SQN might have be seen by the input gate during the announcement
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java [hunk 7]
```diff
@@ -229,11 +229,8 @@ public void run() {
 
 		numBytesIn.inc(buffer.getSize());
 		numBuffersIn.inc();
-		if (buffer.getDataType().hasPriority()) {
-			channelStatePersister.checkForBarrier(buffer);
-		} else {
-			channelStatePersister.maybePersist(buffer);
-		}
+		channelStatePersister.checkForBarrier(buffer);
+		channelStatePersister.maybePersist(buffer);
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java:233]: I've added `LocalInputChannelTest.testNoDataPersistedAfterReceivingAlignedBarrier`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 8]
```diff
@@ -451,11 +451,17 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
...
 						firstPriorityEvent = addPriorityBuffer(announce(sequenceBuffer));
 					}
 				}
+				channelStatePersister.checkForBarrier(sequenceBuffer.buffer).ifPresent(id -> {
+					// checkpoint was not yet started by task thread,
+					// so remember the numbers of buffers to spill for the time when it will be started
+					lastBarrierSequenceNumber = sequenceBuffer.sequenceNumber;
+					lastBarrierId = id;
+				});
+				channelStatePersister.maybePersist(buffer);
```
[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: I didn't add a unit test as after the other fixes in master (#14052) this change is not strictly necessary
(though I think it's still less error-prone to not update SQN unnecessarily).

**Comment by rkhachatryan** [comment]
I've updated the PR and addressed your feedback @pnowojski, please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 9]
```diff
@@ -451,11 +451,20 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
...
+				channelStatePersister
+					.checkForBarrier(sequenceBuffer.buffer)
+					.filter(id -> id > lastBarrierId)
+					.ifPresent(id -> {
+						// checkpoint was not yet started by task thread,
+						// so remember the numbers of buffers to spill for the time when it will be started
+						lastBarrierId = id;
+						lastBarrierSequenceNumber = sequenceBuffer.sequenceNumber;
+					});
+				channelStatePersister.maybePersist(buffer);
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:467]: And what about test coverage for  those changes?

Side question, shouldn't those two fixes be separate commits?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java [hunk 10]
```diff
@@ -68,6 +68,22 @@
  */
 public class AlternatingControllerTest {
 
+	/**
+	 * Upon subsuming (or canceling) a checkpoint, channels should be notified regardless of whether UC controller is
+	 * currently being used or not. Otherwise, channels may capture in-flight buffers from an older checkpoint.
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java:75]: > Otherwise, channels may capture in-flight buffers from an older checkpoint

Is this test actually checking for that? I do not see any buffer that would belong to an older checkpoint?

(I think I still do not understand this fix)

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java [hunk 11]
```diff
@@ -84,6 +84,24 @@ public void testChannelResetOnNewBarrier() throws Exception {
 		assertFalse(stateWriter.getAddedInput().isEmpty());
 	}
 
+	/**
+	 * If a checkpoint announcement was processed and then UC-barrier arrives (from the upstream)
+	 * then it should be processed by the UC controller.
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java:91]:
```
	 * If a checkpoint announcement was processed and then UC-barrier arrives (from the upstream)
	 * then it should be processed by the UC controller.
```
->
```
	 * If a checkpoint announcement was processed from one channel and then UC-barrier arrives 
	 * on another channel, this UC barrier should be processed by the UC controller.
```
?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java [hunk 12]
```diff
@@ -84,6 +84,24 @@ public void testChannelResetOnNewBarrier() throws Exception {
...
+		inputGate.setInputChannels(new TestInputChannel(inputGate, 0), new TestInputChannel(inputGate, 1));
+		ValidatingCheckpointHandler target = new ValidatingCheckpointHandler();
+		SingleCheckpointBarrierHandler barrierHandler = barrierHandler(inputGate, target);
+		CheckpointedInputGate gate = buildGate(target, 2);
+
+		CheckpointBarrier aligned = new CheckpointBarrier(1, System.currentTimeMillis(), alignedWithTimeout(getDefault(), Integer.MAX_VALUE));
+
+		send(toBuffer(new EventAnnouncement(aligned, 0), true), 0, gate); // process announcement but not the barrier
+		send(toBuffer(aligned.asUnaligned(), true), 1, gate); // pretend it came from upstream before the first (AC) barrier was picked up
+	}
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java:105]: Aren't we missing some assertion?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java (hunk 2 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java:117]: I still don't get this fix.
> But there is a preProcessFirstBarrier before the switch in barrierReceived.

Yes, but my intention was that in that case, `preProcessFirstBarrier` would be called on the AlignedController. Next on the first barrier, we would switch to unaligned controller, which is doing this:
```
		// alignedController might has already processed some barriers, so "migrate"/forward those calls to unalignedController.
		unalignedController.preProcessFirstBarrier(channelInfo, barrier);
		for (InputChannelInfo blockedChannel : blockedChannels) {
			unalignedController.barrierReceived(blockedChannel, barrier);
		}
```
so `preProcessFirstBarrier` would be eventually called on the unaligned controller.

**Review by pnowojski — CHANGES_REQUESTED** [review]
Quite a bit of fixes/commits are still missing unit tests. By
> (as in other places: a unit test would be helpful)

I meant if you could add unit tests for all the bug fixes that you implemented.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java (hunk 9 above)
Comment (rkhachatryan) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:467]: > Side question, shouldn't those two fixes be separate commits?

They were in the beginning, but it seemed to me just way too many commits :) So I tried to group some related changes. I can extract it if you prefer.

> And what about test coverage for those changes?

As [noted](https://github.com/apache/flink/pull/14057#discussion_r543678627) above
>  I didn't add a unit test as after the other fixes in master (#14052) this change is not strictly necessary
> (though I think it's still less error-prone to not update SQN unnecessarily).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java (hunk 10 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java:75]: Sorry, my bad, I didn't update the comment after updating the test. 
It checks whether the buffer (on the other channel) was captured or not. I'll update the comment.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java (hunk 12 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingControllerTest.java:105]: The expectation is that it just won't fail (without the fix, it will).

**Review by pnowojski — APPROVED** [review]

**Comment by pnowojski** [comment]
@flinkbot run azure
